mobile/src/main/java/ch/epfl/sdp/mobile/ui/speech_recognition/ChessSpeechRecognitionFilter.kt

```diff
@@ -0,0 +1,210 @@
+  }
+
+  /**
+   * Makes a decision of choosing a conform by relying on confidence scores
+   * @param speeches list of tokenized speech recognizer results
+   * @param confidencesScores confidence scores for each speech result in speeches
+   * @param detectionBlock a detection method for a certain type of chess keyword (ChessPiece,
+   * Number, Letter)
+   * @return a result with confidence or empty if could not determine one
+   */
+  private fun detectFromSpeeches(
+      speeches: List<List<Word>>,
+      confidencesScores: List<Double>,
+      detectionBlock: (List<Word>) -> String,
+  ): String {
+    val candidates = mutableMapOf<String, Double>()
+
+    speeches.forEachIndexed { index, speech ->
+      val candidate = detectionBlock(speech)
+      if (candidate.isNotEmpty()) {
+        candidates[candidate] = confidencesScores[index]
+      }
+    }
+
+    return bestCandidate(candidates)
+  }
+
+  /**
+   * Searches for any matches of for a list word given a dictionary of chess keywords
+   * @param words a single tokenized speech as a list of words
+   * @param dictionary a mapping from chess keyword to its predefined possible homonyms
+   * @return a detected key word from dictionary if any, an empty string otherwise
+   */
+  private fun detect(words: List<Word>, dictionary: Map<String, List<Word>>): String {
+    var result = ""
+    loop@ for (word in words) {
+      for (entry in dictionary) {
+        if (entry.value.any { it == word }) {
+          result = entry.key
+          break@loop
+        }
+      }
+    }
+    return result
+  }
+  /**
+   * Tokenizes speeches and put words to lower case
+   * @param speeches list of the speech recognizer results
+   * @return tokenized lists of speeches ready to be filtered
+   */
+  private fun tokenizeSpeeches(speeches: List<Speech>): List<List<Word>> {
+    return speeches.map { speeche -> speeche.split(" ").map { word -> word.lowercase() } }
+  }
+
+  override fun filterWords(speeches: List<String>, confidencesScores: List<Double>): ChessWord {
+
+    val tokenizedSpeeches = tokenizeSpeeches(speeches)
+
+    // Apply piece speech rules
+    val pieceFromRule = applyRules(tokenizedSpeeches, confidencesScores)
+
+    // Filter using dictionary
+    val pieceFromDictionary =
+        detectFromSpeeches(tokenizedSpeeches, confidencesScores, this::detectPieceType)
+
+    // Prioritizes dictionary filtering over rule
+    val piece =
+        when (pieceFromDictionary.isNotEmpty()) {
+          true -> pieceFromDictionary
+          else ->
+              when (pieceFromRule.isNotEmpty()) {
+                true -> pieceFromRule
+                else -> "!!"
+              }
+        }
+
+    // Apply filters for letter and number
+    val letter =
+        detectFromSpeeches(tokenizedSpeeches, confidencesScores, this::detectLetter).ifEmpty { "@" }
+    val number =
+        detectFromSpeeches(tokenizedSpeeches, confidencesScores, this::detectNumber).ifEmpty { "$" }
```

Comment:
I don't think it's a good idea to have these magic number to define that we don't have a correspondence.

I think that if one element didn't match, we just return an empty string

Reply:
My idea is to at least get some info from the first user unput and then retry. For example if the speech filter recognizer the chess piece and number but not the letter, we retry. 
If the second time, however, we could only recognize the letter but neither the chess piece nor the letter. We could then use the prior info.

Reply:
This approach didn't seem natural for me. As a user, it's more natural for me to say the whole input again.

I propose to let it as is for the moment and discuss this in the next meeting. 

We can change it according what we decide during the meeting in another PR (It's a small change)

Reply:
Seems good ! I also meant that the user will say the whole input again and hopefully the second time we'll be able to detect the missed word-s and reconstruct the whole move (assuming the player didn't change the move)